integration/pkcs11/pkcs11_test.go

```diff
@@ -0,0 +1,420 @@
+	Expect(err).NotTo(HaveOccurred())
+
+	By("Updating the peer signcerts")
+
+	newOrdererPemCert := buildCert(network, caBytes, orgCAPath, peerCSR, peerSerial, peerPubKey)
+
+	updateMSPFolder(network.PeerLocalMSPDir(peer), fmt.Sprintf("peer.%s.example.com-cert.pem", orgName), newOrdererPemCert)
+
+	By("Updating the peer admin user signcerts")
+	newAdminPemCert := buildCert(network, caBytes, orgCAPath, adminCSR, adminSerial, adminPubKey)
+
+	orgAdminMSPPath := network.PeerUserMSPDir(peer, "Admin")
+	updateMSPFolder(orgAdminMSPPath, fmt.Sprintf("Admin@%s.example.com-cert.pem", orgName), newAdminPemCert)
+
+}
+
+func configureOrdererPKCS11(ctx *pkcs11.Ctx, sess pkcs11.SessionHandle, network *nwo.Network) {
+
+	orderer := network.Orderer("orderer")
+	orgName := "orderer"
+
+	ordererPubKey, ordererCSR, ordererSerial := createCSR(ctx, sess, orgName, "orderer")
+	adminPubKey, adminCSR, adminSerial := createCSR(ctx, sess, orgName, "admin")
+
+	// Retrieves org CA cert
+	orgCAPath := network.OrdererOrgCADir()
+	caBytes, err := ioutil.ReadFile(filepath.Join(orgCAPath, "ca.example.com-cert.pem"))
+	Expect(err).NotTo(HaveOccurred())
+
+	By("Updating the orderer signcerts")
+
+	newOrdererPemCert := buildCert(network, caBytes, orgCAPath, ordererCSR, ordererSerial, ordererPubKey)
+
+	updateMSPFolder(network.OrdererLocalMSPDir(orderer), "orderer.example.com-cert.pem", newOrdererPemCert)
+
+	By("Updating the orderer admin user signcerts")
+	newAdminPemCert := buildCert(network, caBytes, orgCAPath, adminCSR, adminSerial, adminPubKey)
+
+	orgAdminMSPPath := network.OrdererUserMSPDir(orderer, "Admin")
+	updateMSPFolder(orgAdminMSPPath, "Admin@example.com-cert.pem", newAdminPemCert)
+
+}
+
+// Creates pkcs11 context and session
+func setupPKCS11Ctx(lib, label, pin string) (*pkcs11.Ctx, pkcs11.SessionHandle, error) {
+
+	ctx := pkcs11.New(lib)
+
+	ctx.Initialize()
+
+	slot, err := findPKCS11Slot(ctx, label)
+	Expect(err).ToNot(HaveOccurred())
+
+	sess, err := ctx.OpenSession(slot, pkcs11.CKF_SERIAL_SESSION|pkcs11.CKF_RW_SESSION)
+	Expect(err).ToNot(HaveOccurred())
+
+	// Login
+	err = ctx.Login(sess, pkcs11.CKU_USER, pin)
+	Expect(err).NotTo(HaveOccurred())
+
+	return ctx, sess, err
+
+}
+
+// Identifies pkcs11 slot using specified label
+func findPKCS11Slot(ctx *pkcs11.Ctx, label string) (uint, error) {
+
+	var (
+		slot      uint
+		foundSlot bool
+	)
+
+	slots, err := ctx.GetSlotList(true)
+	Expect(err).ToNot(HaveOccurred())
+
+	for _, s := range slots {
+		tokInfo, err := ctx.GetTokenInfo(s)
+		Expect(err).ToNot(HaveOccurred())
+
+		if tokInfo.Label == label {
+			foundSlot = true
+			slot = s
+			break
+		}
+	}
+
+	if !foundSlot {
+		Fail("Could not find slot with label")
+		return 0, err
+	}
+
+	return slot, nil
+
+}
+
+// Creates CSR for provided organization and organizational unit
+func createCSR(ctx *pkcs11.Ctx, sess pkcs11.SessionHandle, org, ou string) (*ecdsa.PublicKey, *x509.CertificateRequest, *big.Int) {
```

Comment:
We have multiple copies of a lot of this code floating around. We're at the point where we should consider consolidating them. I know there are versions in common and in comm.